src/Features/Core/Portable/MoveToNamespace/AbstractMoveToNamespaceService.cs

```diff
@@ -0,0 +1,224 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ChangeNamespace;
+using Microsoft.CodeAnalysis.CodeRefactorings.MoveType;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.LanguageServices;
+
+namespace Microsoft.CodeAnalysis.MoveToNamespace
+{
+    internal interface IMoveToNamespaceService : ILanguageService
+    {
+        Task<ImmutableArray<AbstractMoveToNamespaceCodeAction>> GetCodeActionsAsync(Document document, TextSpan span, CancellationToken cancellationToken);
+        Task<MoveToNamespaceAnalysisResult> AnalyzeTypeAtPositionAsync(Document document, int position, CancellationToken cancellationToken);
+        Task<MoveToNamespaceResult> MoveToNamespaceAsync(MoveToNamespaceAnalysisResult analysisResult, string targetNamespace, CancellationToken cancellationToken);
+        MoveToNamespaceOptionsResult GetChangeNamespaceOptions(Document document, string defaultNamespace, ImmutableArray<string> namespaces);
+    }
+
+    internal abstract class AbstractMoveToNamespaceService<TNamespaceDeclarationSyntax, TNamedTypeDeclarationSyntax>
+        : IMoveToNamespaceService
+        where TNamespaceDeclarationSyntax : SyntaxNode
+        where TNamedTypeDeclarationSyntax : SyntaxNode
+
+    {
+        protected abstract string GetNamespaceName(TNamespaceDeclarationSyntax syntax);
+        protected abstract string GetNamespaceName(TNamedTypeDeclarationSyntax syntax);
+
+        public async Task<ImmutableArray<AbstractMoveToNamespaceCodeAction>> GetCodeActionsAsync(
+            Document document,
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var typeAnalysisResult = await AnalyzeTypeAtPositionAsync(document, span.Start, cancellationToken).ConfigureAwait(false);
+
+            if (typeAnalysisResult.CanPerform)
+            {
+                return ImmutableArray.Create(AbstractMoveToNamespaceCodeAction.Generate(this, typeAnalysisResult));
+            }
+
+            return ImmutableArray<AbstractMoveToNamespaceCodeAction>.Empty;
+        }
+
+        public async Task<MoveToNamespaceAnalysisResult> AnalyzeTypeAtPositionAsync(
```

Comment:
Discussed offline
- Remove fallthrough stuff by having two helper methods, one for invocation on namespace and one for types, and then we can talk about each case independently.